When printing groups in the REPL, including information about their generators (and for fp groups, also relations)

fingolfin opened the pull request:
This is an experiment to see what happens if we print the generators of groups, at least in "detailed" mode. I find it quite useful in some cases, but in others it is annoying. Would be interesting to know what others think about this.

lgoettgens commented on src/Groups/directproducts.jl:
```diff
@@ -542,15 +585,24 @@ Return `H`, where `W` is the wreath product of `G` and `H`.
 julia> W = wreath_product(G,H)
 <group of size 18 with 2 generators>
+with 2 generators
+  WreathProductElement(f1,<identity> of ...,())
```
This needs some change to be useful.

fingolfin replied:
I think it is already useful (and compared to not showing anything, conveys infinitely more information). But of course it could be made even more useful :-)

thofma commented:
I think this should have been the default in the first place.

Edit: When do you find it annoying?

fingolfin commented:
@thofma I find it annoying when I have thirty 200x200 matrices as generators (this is not hypothetical, by the way...). So I'd really like a way to limit it based on the screen size...

Also, the generators of subgroups often are determined by randomized algorithms, which can make this kind of printing a PITA when used in doctests (in GAP frequently minor changes to algorithms require annoying changes to doctests -- even with machinery as we have it to update them semi-automatically, it's often a bother). Of course doctest filters may be able to help with that if we encounter it.

I also don't find it very helpful for pc groups where the are, say, 10 generators named `f1` to `f10`. Perhaps I could add code that checks whether the names of the all the generators are short and then print them in a single line or so.

But none of these are insurmountable! (and I would not have made this PR if I was fundamentally opposed to this change, after all ;-)

fingolfin commented on docs/src/InvariantTheory/finite_groups.md:
```diff
@@ -91,6 +91,14 @@ julia> M2 = matrix(K, [1 0 0; 0 a 0; 0 0 -a-1])
+  [0   a        0]
+  [0   0   -a - 1]
 
 julia> IR = invariant_ring(G)
```
The printing of the invariant ring should also be adjusted, I just did not yet get to that (CC @joschmitt)

fingolfin commented on docs/src/InvariantTheory/finite_groups.md:
```diff
@@ -91,6 +91,14 @@ julia> M2 = matrix(K, [1 0 0; 0 a 0; 0 0 -a-1])
 julia> G = matrix_group(M1, M2)
 Matrix group of degree 3
   over cyclotomic field of order 3
+with 2 generators
+  [0   0   1]
+  [1   0   0]
+  [0   1   0]
+
+  [1   0        0]
+  [0   a        0]
+  [0   0   -a - 1]
```
Here it would be nice if there was a hint somewhere that `a` is the name of the generator of the cyclotomic field.

But I notice that the cyclotomic field has no option for printing that. Indeed, even its detailed printing omits the name of its generator:
```
julia> K, a = cyclotomic_field(3, "a")
(Cyclotomic field of order 3, a)

julia> K
Number field with defining polynomial _$^2 + _$ + 1
  over rational field

julia> gen(K)
a
```

In fact I am surprised to see that the detailed printing drops the hint that this is a cyclotomic field (I mean, of course one can see this from the defining polynomial, but in larger degrees I would not be able to recognize these reliably).

fingolfin replied:
Filed under https://github.com/Nemocas/Nemo.jl/issues/2105

fingolfin commented:
@ThomasBreuer do you have any thoughts on this, pro or contra?

thofma commented:
I agree that for some complicated coefficient fields, the printing might not convey all the information, but I think this is outweighed by the advantages.

ThomasBreuer reviewed (commented):
I think that printing full information about group order and generators is meaningful at most for small examples.

For example, try `atlas_group(MatrixGroup, "J3")`: This is a group with 2 generators in dimension 80 over GF(2), probably nobody wants to see such matrices, except if one asks explicitly to look at them.
Or try `sylow_subgroup(symmetric_group(100), 2)[1]`: This group has 97 generators, almost nobody wants to see them.
(And then replace degree 100 by 1000 and see what happens.)

I think these examples are not artificially large, already the 10 by 10 matrices generating the group M11 (that occur in one of the examples) are not interesting enough to be shown if one just wants to work with the group.
Even beginners working with small groups will deal with such large permutations and matrices -- think of regular actions.

Let us compare the situation with that in GAP, where we had similar problems.

An answer was to distinguish between `View` (short output, not nec. GAP-readable) and `Print` (longer output, preferably GAP-readable); the main idea was that in the case of long output, full information shall be given only if the user asks for it, by calling `Print`.
And `View` tries to abbreviate permutations of large degree by showing only the first cycles and then `( [...] )`; of course this does not help if the permutation consists only of one cycle.

The current situation in GAP may still be regarded as not convincing.
I remember of a discussion in which a user argued that in the case of `View` (that is, one has not explicitly asked for more output, presumably one just forgot to type the double semicolon that suppresses the output) the output should always be restricted to the first n lines and the last n lines, with `...` in the middle, where n should be about 5.

thofma commented:
If one does not care for the output in the first place, one can just use a `;`.

ThomasBreuer commented:
@thofma

> If one does not care for the output in the first place, one can just use a ;.

This was one of the points in the abovementioned discussion in the GAP Forum: The user in question did not agree with the argument that typing one more semicolon would save him from many pages of annoying output. He wrote that this argument is the same as a statement somewhere in the fineprint that all one has to do in order to prevent the computer from blowing up is to push the "Dont't blow up" button before switching the machine on.

I understand this comparison as the verdict that a system should not only allow one to avoid obviously useless output but that avoiding such output should be the default.

thofma commented:
A system has a default way of printing and there is an easy way to suppress it. Doesn't sound so complicated to me. A system should not try to decide what  "useless" is. Why not just print "group"? Obviously in my application, this is all I need to know.

ThomasBreuer commented:
@thofma 
What about Julia's default to show vectors by restricting them to one screen?
For example, `gens(sylow_subgroup(symmetric_group(100), 2)[1]` is by default printed like this, the first entries followed by `⋮` followed by the last entries. This is still more than the mentioned GAP user wanted to see, but addresses his complaint to some extent.

lgoettgens commented:
In my opinion, any reasonable "default" of printing should print at most one screen full of stuff. So in this sense, I really like the julia way of printing vectors.

thofma commented:
The problem with `⋮` is that it is almost impossible to get the full information. Assume my group has too many generators (say 10) to fit into my terminal. With the `⋮` idea, we would only print 5 of them. How do I see all of them? If I do `gens(G)`, it is a julia vector and again I have the `⋮` and not the full information. And now? I can do `print`, which is even less readable. Now what? I just want to see what all the elements are without writing a loop that prints everything. How hard can this be?

This `⋮` is all cute and nice and fancy, but too much magic which makes it frustrating to use.

ThomasBreuer commented:
@thofma
> The problem with ⋮ is that it is almost impossible to get the full information.

Good point.
However, I would say that this shows a missing functionality concerning (among others) vectors in Julia.
When I want to show a group `G` then usually I do not want to see all generators, but when I want to show `gens(G)` then this is in general the case.
Thus there should be a function that really shows the whole vector.

simonbrandhorst commented:
I am in favor of the changes. For matrix groups of large degree (say >= 10) I would not print the matrices.

fingolfin commented:
Note to self: also add suitable code to prevent lower casing for `SL` etc., as in this example:
```
julia> G = SL(2,5)
SL(2,5)

julia> T
Character table of sL(2,5)
...
```
Note the `sL` instead of `SL`

fingolfin commented:
I will try to resurrect this and based on triage discussion make it only print stuff that fits, with "..." if there are too many generators / too big ones, etc.

fingolfin commented on docs/src/Groups/pcgroup.md:
```diff
@@ -93,11 +96,19 @@ and let Oscar compute a pc presentation for it.
 Sym(4)
+with 2 generators
+  (1,2,3,4)
+  (1,2)
```
Perhaps for `Sym(n)` we should omit the generators? Might still be useful to know how many generators there are... Perhaps print as "Symmetric group with 2 generator"?

fingolfin replied:
I did this now as an experiment, and also move the "with 2 generators" onto the same line (for permutation groups only! for pc-groups and matrix groups it remains as it is).

fingolfin commented on src/Combinatorics/Graphs/functions.jl:
```diff
@@ -717,6 +717,10 @@ julia> g = complete_graph(4);
 
 julia> automorphism_group(g)
 Permutation group of degree 4
+with 3 generators
```
Would it be better to have this all in one line?
```suggestion
Permutation group of degree 4 with 3 generators
```

simonbrandhorst replied:
Maybe o.k. for permutation groups, but for more nested constructs it will be easier to have multiple lines 
e.g. the following seem bad to me:
`automorphism group of permutation group of degree 4 with 3 generators with 4 generators`
or even worse, when it is not clear at all where the generators belong:
`automorphism group of permutation group of degree 4 with 4 generators`

The following seems more structured to me:
```
automorphism group 
  of permutation group of degree 4 with 3 generators 
with 2 generators`
  f1 
  f2
```

fingolfin commented on src/Groups/abelian_aut.jl:
```diff
@@ -432,7 +435,12 @@ Gram matrix quadratic form:
-Group of isometries of Finite quadratic module: Z/15 -> Q/2Z generated by 2 elements
+Group of isometries of finite quadratic module: Z/15 -> Q/2Z with 2 generators
+with 2 generators
+  Isometry of finite quadratic module: Z/15 -> Q/2Z defined by
```
oops, double "with 2 generators"

fingolfin commented on src/Groups/directproducts.jl:
```diff
@@ -632,7 +702,17 @@ function canonical_injections(W::WreathProductGroup)
+    io = pretty(io)
+    println(io, "Wreath product of", Indent())
+    #print(io, A.G)
+    print(io, "FIXME")
```
FIXME

lgoettgens commented:
The change in handling of how symmetric and alternating groups are printed has an effect on groups that were not created using the constructors, e.g. `perm_group(GAP.Globals.SymmetricGroup(3))` used to print as `Sym(3)` but no longer does. I am not objecting to this change, just wanting to point this out

fingolfin commented on experimental/GaloisGrp/src/Solve.jl:
```diff
@@ -295,8 +295,8 @@ julia> G, C = galois_group(x^3-3*x+17)
- Sym(3)
- Alt(3)
+ Permutation group of degree 3 and order 6
+ Permutation group of degree 3 and order 3
```
I am unhappy about how my experiment with the printing for symmetric and alternating groups works out here. I'd rather go back to how it was... or perhaps this:
```suggestion
 Symmetric group of degree 3 and order 6
 Alternating group of degree 3 and order 3
```

fingolfin commented on src/Groups/directproducts.jl:
```diff
@@ -168,16 +170,16 @@ julia> K = symmetric_group(2)
-Direct product of
- Sym(3)
- Sym(2)
+Sym(3) x Sym(2)
```
Here I am also torn: perhaps the original output should be kept, and `Sym(3) x Sym(2)` just used for oneline resp. terse printing?

fingolfin commented on src/Groups/libraries/atlasgroups.jl:
```diff
@@ -109,21 +128,43 @@ julia> g = atlas_group("M11");  # Mathieu group M11
+  (1,6,10,7,11,3,9,2)(4,5)
 
 julia> h3, emb = atlas_subgroup(MatrixGroup, "M11", 1 );  h3
 Matrix group of degree 10
   over prime field of characteristic 2
+with 2 generators
+  [0   1   0   1   0   1   0   0   0   0]
+  [0   0   1   0   0   0   0   0   0   0]
+  [0   1   0   0   0   0   0   0   0   0]
+  [1   1   1   1   1   0   0   0   0   0]
+  [1   1   0   1   1   1   0   0   0   0]
+  [0   1   0   1   1   0   0   0   0   0]
+  [0   0   1   1   1   1   1   0   0   0]
+  [0   0   0   1   1   0   0   1   1   0]
+  [0   0   1   1   1   1   0   0   1   0]
+  [0   0   1   0   0   1   0   0   1   1]
+
+  ⋮
 
```
Here we can see what happens if the generators of a matrix group "do not fit": I am inserting a vertical dots (unicode). But this should be changed if unicode is not allowed -- but to what? Perhaps just horizontal `. . . ` or `...`?
```suggestion
Matrix group of degree 10
  over prime field of characteristic 2
with 2 generators
  [0   1   0   1   0   1   0   0   0   0]
  [0   0   1   0   0   0   0   0   0   0]
  [0   1   0   0   0   0   0   0   0   0]
  [1   1   1   1   1   0   0   0   0   0]
  [1   1   0   1   1   1   0   0   0   0]
  [0   1   0   1   1   0   0   0   0   0]
  [0   0   1   1   1   1   1   0   0   0]
  [0   0   0   1   1   0   0   1   1   0]
  [0   0   1   1   1   1   0   0   1   0]
  [0   0   1   0   0   1   0   0   1   1]

  ...

```

thofma replied:
Both are equally bad, so just go with the vertical dots. Then we are consistent with the bad solution in julia itself.

Semi-serious suggestion: print something like
```
[1 generator omitted]
```

fingolfin commented on docs/src/Groups/pcgroup.md:
````diff
@@ -92,28 +92,30 @@ and let Oscar compute a pc presentation for it.
 ```jldoctest
 julia> dihedral_group(8)
-Pc group of order 8
+Pc group of order 8 with 3 generators [f1, f2, f3]
````
I experimented further with printing the generators of (sub) fp-groups and (sub) pc-groups and arrived at this, inspired by how we print multivariate polynomial rings.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -588,13 +649,13 @@ Return whether generators for the group `G` are known.
 
 julia> H = derived_subgroup(F)[1]
-Free group
+Free group of unknown rank
```
For free groups we already printed the rank if it was known, but nothing if it was unknown. Now I print explicitly "of unknown rank". Why is visible in this example:
```
julia> f = free_group(4)
Free group of rank 4 with 4 generators [f1, f2, f3, f4]

julia> h = sub(f, [f[1],f[3], f[3]*f[1]])[1]
Free group of unknown rank with 3 generators [f1, f3, f3*f1]
```
So this subgroup is a free group of rank 2, but it was printed as `Free group with 3 generators` which I think one can very very easily mistake for implying that the rank is 3.

Or perhaps it is just me who is not careful enough... but I really think this is beneficial, and the downsides seem to be minimal?

fingolfin commented on src/Groups/directproducts.jl:
```diff
@@ -17,15 +17,20 @@ vectors of the embeddings (resp. projections) of the direct product `G`.
+  (1,2,3)
+  (1,2)
+  (4,5)
 
```
Not sure I like this. Perhaps should revert to the old form, and only keep the new one for oneline / terse printing

lgoettgens replied:
I agree. For detailed this is a lot harder to read

fingolfin commented on src/Groups/directproducts.jl:
```diff
@@ -162,22 +167,25 @@ It is not defined for proper subgroups of direct products.
-   Sym(3)
-   Sym(2)
+  from symmetric group of degree 3 and order 6
+  to symmetric group of degree 3 and order 6 x Symmetric group of degree 2 and order 2
```
Also not happy about this. I think now better to keep it as it was, or change to one of these
```suggestion
  to Sym(3) x Sym(2)
```
or
```suggestion
  to direct product of
    symmetric group of degree 3 and order 6
    symmetric group of degree 2 and order 2
```

fingolfin commented on src/Groups/directproducts.jl:
```diff
@@ -355,27 +367,41 @@ where `f` is a group homomorphism from `H` to the automorphism group of `N`.
 julia> derived_subgroup(G)
-(Group([ f4 ]), Hom: group([ f4 ]) -> semidirectProduct( <pc group of size 8 with 3 generators> , <pc group of size 2 with 1 generator> ))
+(Subgroup of semidirect product
+  pc group of order 8 : pc group of order 2, Hom: subgroup of semidirect product -> semidirect product)
```
The line break should not be in there

lgoettgens commented on experimental/GaloisGrp/src/Solve.jl:
```diff
@@ -291,12 +291,12 @@ one, compute the corresponding subfields as a tower.
 3-element Vector{PermGroup}:
- Sym(3)
- Alt(3)
+ Symmetric group of degree 3 and order 6
+ Alternating group of degree 3 and order 3
```
Do we really need Sym and Alt groups printing their order? For general PermGroups, I see how that is useful, but for these two "groups with names" IMO this clutters the output more than it helps

fingolfin replied:
fine by me to drop it. It gets large very quickly anyway.

lgoettgens commented on src/Combinatorics/Matroids/matroids.jl:
```diff
@@ -1010,7 +1010,10 @@ julia> M = uniform_matroid(2, 4)
+Permutation group of degree 4 with 3 generators
+  (3,4)
+  (1,2)
+  (2,3)
```
Not objecting to this change, but other types don't print the number of gens in the cases where the gens are listed, e.g.:
```julia
julia> R, x = polynomial_ring(QQ, 10);

julia> ideal(R, [x[1], x[3]^3*x[9]])
Ideal generated by
  x1
  x3^3*x9
```
I am not sure what of the two I prefer

fingolfin replied:
Though polynomial rings do:
```
julia> R
Multivariate polynomial ring in 10 variables x1, x2, x3, x4, ..., x10
  over rational field
```
But still, you raise a valid point. Personally I like to see the number of generators, at least in the example of a permutation group, because there it is not always clear on a glance. Of course if you have variables named x1 to x10 the number of variables is kinda redundant.

Likewise for e.g. an ideal which can have arbitrary polynomials as generators, it seems useful to also include the number, so I'd be tempted to say the ideal printing in your example should be changed :-).

But in the end this is not a terribly import point to me, I am happy to keep it or change it for the groups either way. Perhaps we'll see if others have stronger opinions on this ;-)


On a tangent, I like how the polynomial ring abbreviates its generator list, perhaps that's something to consider for fp and pc groups as well... Then again I am not sure they'll be produced with too many generators often... *shrug*.

fingolfin replied:
Thinking some more, perhaps labeling the generators would actually be useful, as that helps me to later reference them (e.g. as `G[5]` etc.).
```
Permutation group of degree 4 with 3 generators
  1: (3,4)
  2: (1,2)
  3: (2,3)
```

lgoettgens commented on src/Groups/directproducts.jl:
```diff
@@ -504,13 +530,18 @@ W(g_1,...,g_n, h).
 julia> W = wreath_product(G,H)
-<group of size 18 with 2 generators>
+Wreath product of
+  pc group of order 3 wr symmetric group of degree 2 and order 2
```
From a linguistic POV I would say that after `Wreath product of` there should either be a list of things or something like `A in B` or so, but not `A wr B`, which is already the complete notation for the wreath product

lgoettgens commented on src/Groups/directproducts.jl:
```diff
@@ -504,13 +530,18 @@ W(g_1,...,g_n, h).
+  pc group of order 3 wr symmetric group of degree 2 and order 2
+with 2 generators
+  WreathProductElement(f1,<identity> of ...,())
+  WreathProductElement(<identity> of ...,<identity> of ...,(1,2))
```
I really have no idea what this output should convey. But I think this is one of the types you skipped for now, so please ignore this comment until you tackle this

lgoettgens commented on src/Groups/libraries/atlasgroups.jl:
```diff
@@ -13,14 +13,28 @@ for `name` is available, and `MatrixGroup` otherwise.
+  [0   0   1   0]
+  [0   1   0   0]
+  [1   1   1   1]
+
```
Have you checked that things like `doctest_fix` work as expected if the printing of some object contains empty lines? In particular I would be interested in if this manages to fix typos in the second generator' matrix (aka after the empty line)

lgoettgens reviewed (commented):
some random comments with personal opinions

fingolfin committed:
Regenerate doctests

fingolfin committed:
Revise detailed printing for group

- show information about group generators
- change printing of symmetric and alternating groups
- improve printing of various group products (direct, semidirect, wreath)

fingolfin committed:
Adjust some tests manually

fingolfin commented:
So for permutation and symmetric groups there clearly is a TODO, given this:
```
julia> symmetric_group(500)
Symmetric group of degree 500 and order 1220136825991110068701238785423046926253574342803192842192413588385845373153881997605496447502203281863013616477148203584163378722078177200480785205159329285477907571939330603772960859086270429174547882424912726344305670173270769461062802310452644218878789465754777149863494367781037644274033827365397471386477878495438489595537537990423241061271326984327745715546309977202781014561081188373709531016356324432987029563896628911658974769572087926928871281780070265174507768410719624390394322536422605234945850129918571501248706961568141625359056693423813008856249246891564126775654481886506593847951775360894005745238940335798476363944905313062323749066445048824665075946735862074637925184200459369692981022263971952597190945217823331756934581508552332820762820023402626907898342451712006207714640979456116127629145951237229913340169552363850942885592018727433795173014586357570828355780158735432768888680120399882384702151467605445407663535984174430480128938313896881639487469658817504506926365338175055478128640000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 with 2 generators
  (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,
  41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,7
  8,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111
  ,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139
  ,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167
  ,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195
  ,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223
  ,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251
  ,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279
  ,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,304,305,306,307
  ,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335
  ,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363
  ,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391
  ,392,393,394,395,396,397,398,399,400,401,402,403,404,405,406,407,408,409,410,411,412,413,414,415,416,417,418,419
  ,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435,436,437,438,439,440,441,442,443,444,445,446,447
  ,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,474,475
  ,476,477,478,479,480,481,482,483,484,485,486,487,488,489,490,491,492,493,494,495,496,497,498,499,500)
  (1,2)
```

TODO:
1. abbreviate permutations (that likely means we need our own permutation printing code, but that's easy enough)
2. Abbreviate or omit the order in such cases. Several options for that
   1. @lgoettgens suggested just printing it for alternating/symmetric groups
   2. we could also just omit the order if it has more than, say, 20 digits (that's an arbitrary bound
   3. we could also "abbreviate" the order, like GAP does, by only printing the first and last couple digits and an ellipsis in betweeen
   4. or we could use scientific notation and print "of order ~1.22e1134" or so

fingolfin committed:
tweak printing

fingolfin committed:
WIP

thofma commented:
Would be great if we could also tweak the printing of finitely presented groups. Maybe
```
Finitely presented groups with x generators and y relators
```
or even print the list of relators.

fingolfin committed:
Print number of relators for FPGroup

fingolfin committed:
more fpgroup printing tweaks

fingolfin committed:
WIP

fingolfin committed:
Tweak sym/alt grp printing

fingolfin committed:
fixup! Regenerate doctests
